File: Source/Alternet.UI/WxWidgets/Sizer/SizerItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Alternet.Drawing;

namespace Alternet.UI
{
    internal class SizerItem : DisposableObject<IntPtr>, ISizerItem
    {
        public SizerItem(
            Control window,
            int proportion = 0,
            SizerFlag flag = 0,
            int border = 0)
            : this(
                  Native.SizerItem.CreateSizerItem(
                    WxPlatform.WxWidget(window),
                    proportion,
                    (int)flag,
                    border,
                    default), true)
        {
        }

        public SizerItem(Control window, ISizerFlags sizerFlags)
            : this(
                  Native.SizerItem.CreateSizerItem2(
                WxPlatform.WxWidget(window),
                ((SizerFlags)sizerFlags).Handle),
                  true)
        {
        }

        public SizerItem(
            ISizer sizer,
            int proportion = 0,
            SizerFlag flag = 0,
            int border = 0)
            : this(
                  Native.SizerItem.CreateSizerItem3(
                      ((Sizer)sizer).Handle,
                      proportion,
                      (int)flag,
                      border,
                      default), true)
        {
        }

        public SizerItem(ISizer sizer, ISizerFlags sizerFlags)
            : this(Native.SizerItem.CreateSizerItem4(((Sizer)sizer).Handle, ((SizerFlags)sizerFlags).Handle), true)
        {
        }

        public SizerItem(
            int width,
            int height,
            int proportion = 0,
            SizerFlag flag = 0,
            int border = 0)
            : this(
                  Native.SizerItem.CreateSizerItem5(
                    width,
                    height,
                    proportion,
                    (int)flag,
                    border,
                    default), true)
        {
        }

        public SizerItem(int width, int height, ISizerFlags sizerFlags)
            : this(Native.SizerItem.CreateSizerItem6(width, height, ((SizerFlags)sizerFlags).Handle), true)
        {
        }

        public SizerItem()
            : this(Native.SizerItem.CreateSizerItem7(), true)
        {
        }

        public SizerItem(IntPtr handle, bool disposeHandle)
            : base(handle, disposeHandle)
        {
        }

        public bool IsShown
        {
            get
            {
                return Native.SizerItem.IsShown(Handle);
            }
        }

        public SizeI Size
        {
            get
            {
                return Native.SizerItem.GetSize(Handle);
            }
        }

        public SizeI MinSizeWithBorder
        {
            get
            {
                return Native.SizerItem.GetMinSizeWithBorder(Handle);
            }
        }

        public SizeI MaxSize
        {
            get
            {
                return Native.SizerItem.GetMaxSize(Handle);
            }
        }

        public SizeI MaxSizeWithBorder
        {
            get
            {
                return Native.SizerItem.GetMaxSizeWithBorder(Handle);
            }
        }

        public SizeI MinSize
        {
            get
            {
                return Native.SizerItem.GetMinSize(Handle);
            }

            set
            {
                Native.SizerItem.SetMinSize(Handle, value.Width, value.Height);
            }
        }

        public int Id
        {
            get
            {
                return Native.SizerItem.GetId(Handle);
            }

            set
            {
                Native.SizerItem.SetId(Handle, value);
            }
        }

        public bool IsControl
        {
            get
            {
                return Native.SizerItem.IsWindow(Handle);
            }
        }

        public bool IsSizer
        {
            get
            {
                return Native.SizerItem.IsSizer(Handle);
            }
        }

        public bool IsSpacer
        {
            get
            {
                return Native.SizerItem.IsSpacer(Handle);
            }
        }

        public int Proportion
        {
            get
            {
                return Native.SizerItem.GetProportion(Handle);
            }

            set
            {
                Native.SizerItem.SetProportion(Handle, value);
            }
        }

        public SizerFlag Flag
        {
            get
            {
                return (SizerFlag)Native.SizerItem.GetFlag(Handle);
            }

            set
            {
                Native.SizerItem.SetFlag(Handle, (int)value);
            }
        }

        public int Border
        {
            get
            {
                return Native.SizerItem.GetBorder(Handle);
            }

            set
            {
                Native.SizerItem.SetBorder(Handle, value);
            }
        }

        public float Ratio
        {
            get
            {
                return Native.SizerItem.GetRatio(Handle);
            }

            set
            {
                Native.SizerItem.SetRatio2(Handle, value);
            }
        }

        /// <summary>
        /// Gets or sets the userData item attribute.
        /// </summary>
        internal IntPtr UserData
        {
            get
            {
                return Native.SizerItem.GetUserData(Handle);
            }

            set
            {
                Native.SizerItem.SetUserData(Handle, value);
            }
        }

        public void DetachSizer()
        {
            Native.SizerItem.DetachSizer(Handle);
        }

        public void DetachControl()
        {
            Native.SizerItem.DetachWindow(Handle);
        }

        public SizeI CalcMin()
        {
            return Native.SizerItem.CalcMin(Handle);
        }

        public void SetDimension(PointI pos, SizeI size)
        {
            Native.SizerItem.SetDimension(Handle, pos, size);
        }

        public void SetInitSize(int w, int h)
        {
            Native.SizerItem.SetInitSize(Handle, w, h);
        }

        public void SetRatio(int width, int height)
        {
            Native.SizerItem.SetRatio(Handle, width, height);
        }

        public RectI GetRect()
        {
            return Native.SizerItem.GetRect(Handle);
        }

        public ISizer GetSizer()
        {
            var result = Native.SizerItem.GetSizer(Handle);
            return new Sizer(result, false);
        }

        public SizeI GetSpacer()
        {
            return Native.SizerItem.GetSpacer(Handle);
        }

        public void Show(bool show = true)
        {
            Native.SizerItem.Show(Handle, show);
        }

        public PointI GetPosition()
        {
            return Native.SizerItem.GetPosition(Handle);
        }

        public void SetControl(Control control)
        {
            Native.SizerItem.AssignWindow(Handle, WxPlatform.WxWidget(control));
        }

        public void SetSizer(ISizer sizer)
        {
            if(sizer is null)
                Native.SizerItem.AssignSizer(Handle, default);
            else
                Native.SizerItem.AssignSizer(Handle, ((Sizer)sizer).Handle);
        }

        public void SetSpacer(int w, int h)
        {
            Native.SizerItem.AssignSpacer(Handle, w, h);
        }

        /// <summary>
        /// Destroys the control or the controls in a subsizer, depending on the type of item.
        /// </summary>
        internal void DeleteControls()
        {
            Native.SizerItem.DeleteWindows(Handle);
        }

        /// <summary>
        /// If this item is tracking a control then return it; <c>null</c> otherwise.
        /// </summary>
        internal IntPtr GetControl()
        {
            return Native.SizerItem.GetWindow(Handle);
        }

        // Called once the first component of an item has been decided. This is
        // used in algorithms that depend on knowing the size in one direction
        // before the min size in the other direction can be known.
        // Returns true if it made use of the information (and min size was changed).
        internal bool InformFirstDirection(
            int direction,
            int size,
            int availableOtherDir = -1)
        {
            return Native.SizerItem.InformFirstDirection(Handle, direction, size, availableOtherDir);
        }
    }
}

File: Old/RemovedNativeApi/RemovedFromPal/SizerItem.cpp
#include "SizerItem.h"

namespace Alternet::UI
{
    SizerItem::SizerItem()
    {
    }

    SizerItem::~SizerItem()
    {
    }

    void* SizerItem::CreateSizerItem(void* window, int proportion,
        int flag, int border, void* userData)
    {
        return new wxSizerItem((wxWindow*)window, proportion,
            flag, border, (wxObject*)userData);
    }

    void* SizerItem::CreateSizerItem2(void* window, void* sizerFlags)
    {
        return new wxSizerItem((wxWindow*)window, *(wxSizerFlags*)sizerFlags);
    }

    void* SizerItem::CreateSizerItem3(void* sizer, int proportion, int flag, int border, void* userData)
    {
        return new wxSizerItem((wxSizer*)sizer, proportion, flag, border, (wxObject*) userData);
    }

    void* SizerItem::CreateSizerItem4(void* sizer, void* sizerFlags)
    {
        return new wxSizerItem((wxSizer*)sizer, *(wxSizerFlags*)sizerFlags);
    }

    void* SizerItem::CreateSizerItem5(int width, int height, int proportion, int flag,
        int border, void* userData)
    {
        return new wxSizerItem(width, height, proportion, flag,
            border, (wxObject*) userData);
    }

    void* SizerItem::CreateSizerItem6(int width, int height, void* sizerFlags)
    {
        return new wxSizerItem(width, height, *(wxSizerFlags*) sizerFlags);
    }

    void* SizerItem::CreateSizerItem7()
    {
        return new wxSizerItem();
    }

    void SizerItem::DeleteWindows(void* handle)
    {
        ((wxSizerItem*)handle)->DeleteWindows();
    }

    void SizerItem::DetachSizer(void* handle)
    {
        ((wxSizerItem*)handle)->DetachSizer();
    }

    void SizerItem::DetachWindow(void* handle)
    {
        ((wxSizerItem*)handle)->DetachWindow();
    }

    Int32Size SizerItem::GetSize(void* handle)
    {
        return ((wxSizerItem*)handle)->GetSize();
    }

    Int32Size SizerItem::CalcMin(void* handle)
    {
        return ((wxSizerItem*)handle)->CalcMin();
    }

    void SizerItem::SetDimension(void* handle, const Int32Point& pos, const Int32Size& size)
    {
        ((wxSizerItem*)handle)->SetDimension(pos, size);
    }

    Int32Size SizerItem::GetMinSize(void* handle)
    {
        return ((wxSizerItem*)handle)->GetMinSize();
    }

    Int32Size SizerItem::GetMinSizeWithBorder(void* handle)
    {
        return ((wxSizerItem*)handle)->GetMinSizeWithBorder();
    }

    Int32Size SizerItem::GetMaxSize(void* handle)
    {
        return ((wxSizerItem*)handle)->GetMaxSize();
    }

    Int32Size SizerItem::GetMaxSizeWithBorder(void* handle)
    {
        return ((wxSizerItem*)handle)->GetMaxSizeWithBorder();
    }

    void SizerItem::SetMinSize(void* handle, int x, int y)
    {
        return ((wxSizerItem*)handle)->SetMinSize(x, y);
    }

    void SizerItem::SetInitSize(void* handle, int x, int y)
    {
        return ((wxSizerItem*)handle)->SetInitSize(x, y);
    }

    void SizerItem::SetRatio(void* handle, int width, int height)
    {
        return ((wxSizerItem*)handle)->SetRatio(width, height);
    }

    void SizerItem::SetRatio2(void* handle, float ratio)
    {
        return ((wxSizerItem*)handle)->SetRatio(ratio);
    }

    float SizerItem::GetRatio(void* handle)
    {
        return ((wxSizerItem*)handle)->GetRatio();
    }

    Int32Rect SizerItem::GetRect(void* handle)
    {
        return ((wxSizerItem*)handle)->GetRect();
    }

    void SizerItem::SetId(void* handle, int id)
    {
        return ((wxSizerItem*)handle)->SetId(id);
    }

    int SizerItem::GetId(void* handle)
    {
        return ((wxSizerItem*)handle)->GetId();
    }

    bool SizerItem::IsWindow(void* handle)
    {
        return ((wxSizerItem*)handle)->IsWindow();
    }

    bool SizerItem::IsSizer(void* handle)
    {
        return ((wxSizerItem*)handle)->IsSizer();
    }

    bool SizerItem::IsSpacer(void* handle)
    {
        return ((wxSizerItem*)handle)->IsSpacer();
    }

    void SizerItem::SetProportion(void* handle, int proportion)
    {
        return ((wxSizerItem*)handle)->SetProportion(proportion);
    }

    int SizerItem::GetProportion(void* handle)
    {
        return ((wxSizerItem*)handle)->GetProportion();
    }

    void SizerItem::SetFlag(void* handle, int flag)
    {
        return ((wxSizerItem*)handle)->SetFlag(flag);
    }

    int SizerItem::GetFlag(void* handle)
    {
        return ((wxSizerItem*)handle)->GetFlag();
    }

    void SizerItem::SetBorder(void* handle, int border)
    {
        return ((wxSizerItem*)handle)->SetBorder(border);
    }

    int SizerItem::GetBorder(void* handle)
    {
        return ((wxSizerItem*)handle)->GetBorder();
    }

    void* SizerItem::GetWindow(void* handle)
    {
        return ((wxSizerItem*)handle)->GetWindow();
    }

    void* SizerItem::GetSizer(void* handle)
    {
        return ((wxSizerItem*)handle)->GetSizer();
    }

    Int32Size SizerItem::GetSpacer(void* handle)
    {
        return ((wxSizerItem*)handle)->GetSpacer();
    }

    bool SizerItem::IsShown(void* handle)
    {
        return ((wxSizerItem*)handle)->IsShown();
    }

    void SizerItem::Show(void* handle, bool show)
    {
        return ((wxSizerItem*)handle)->Show(show);
    }

    void SizerItem::SetUserData(void* handle, void* userData)
    {
        return ((wxSizerItem*)handle)->SetUserData((wxObject*)userData);
    }

    void* SizerItem::GetUserData(void* handle)
    {
        return ((wxSizerItem*)handle)->GetUserData();
    }

    Int32Point SizerItem::GetPosition(void* handle)
    {
        return ((wxSizerItem*)handle)->GetPosition();
    }

    bool SizerItem::InformFirstDirection(void* handle, int direction, int size, int availableOtherDir)
    {
        return ((wxSizerItem*)handle)->InformFirstDirection(direction, size, availableOtherDir);
    }

    void SizerItem::AssignWindow(void* handle, void* window)
    {
        return ((wxSizerItem*)handle)->AssignWindow((wxWindow*)window);
    }

    void SizerItem::AssignSizer(void* handle, void* sizer)
    {
        return ((wxSizerItem*)handle)->AssignSizer((wxSizer*)sizer);
    }

    void SizerItem::AssignSpacer(void* handle, int w, int h)
    {
        return ((wxSizerItem*)handle)->AssignSpacer(w, h);
    }
}
